// @storybook/components/dist/blocks/PropsTable/PropRow.js
"use strict";

require("core-js/modules/es.symbol");

require("core-js/modules/es.symbol.description");

require("core-js/modules/es.function.name");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PropRow = void 0;

var _react = _interopRequireDefault(require("react"));

var _markdownToJsx = _interopRequireDefault(require("markdown-to-jsx"));

var _theming = require("@storybook/theming");

var _polished = require("polished");

var _PropJsDoc = require("./PropJsDoc");

var _PropValue = require("./PropValue");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var Name = _theming.styled.span({
  fontWeight: 'bold'
});

var Required = _theming.styled.span(function (_ref) {
  var theme = _ref.theme;
  return {
    color: theme.color.negative,
    fontFamily: theme.typography.fonts.mono,
    cursor: 'help'
  };
});

var Type = _theming.styled.div(function (_ref2) {
  var theme = _ref2.theme;
  return {
    color: theme.base === 'light' ? (0, _polished.transparentize)(0.4, theme.color.defaultText) : (0, _polished.transparentize)(0.6, theme.color.defaultText),
    fontFamily: theme.typography.fonts.mono,
    fontSize: "".concat(theme.typography.size.code, "%")
  };
});

var _ref4 =
/*#__PURE__*/
_react["default"].createElement(Required, {
  title: "Required"
}, "*");

var PropRow = function PropRow(_ref3) {
  var _ref3$row = _ref3.row,
      name = _ref3$row.name,
      type = _ref3$row.type,
      required = _ref3$row.required,
      description = _ref3$row.description,
      defaultValue = _ref3$row.defaultValue,
      jsDocTags = _ref3$row.jsDocTags;
  return _react["default"].createElement("tr", null, _react["default"].createElement("td", null, _react["default"].createElement(Name, null, name), required ? _ref4 : null), _react["default"].createElement("td", null, _react["default"].createElement(_markdownToJsx["default"], null, description || ''), _react["default"].createElement(Type, null, _react["default"].createElement(_PropValue.PropValue, {
    value: type
  })), _react["default"].createElement(_PropJsDoc.PropJsDoc, {
    tags: jsDocTags
  })), _react["default"].createElement("td", null, _react["default"].createElement(_PropValue.PropValue, {
    value: defaultValue
  })));
};

exports.PropRow = PropRow;
PropRow.displayName = "PropRow";
